docs/contributing/plugin-tutorial.md

```diff
@@ -29,254 +64,153 @@ Here is a visual statement of what we want to accomplish:
-of the code lens to the LSP client, who will send it back to our command handler when the user activates
-the code lens:
-```haskell
-importCommandId :: CommandId
-importCommandId = "ImportLensCommand"
+`CommandFunction` takes an `ideState` and a JSON-encodable argument.
 
-importLensCommand :: PluginCommand
-importLensCommand =
-    PluginCommand importCommandId "Explicit import command" runImportCommand
+Our handler will ignore the state argument and only use the `WorkspaceEdit` argument.
 
+```haskell
 -- | The type of the parameters accepted by our command
-data ImportCommandParams = ImportCommandParams WorkspaceEdit
-  deriving Generic
+newtype ImportCommandParams = ImportCommandParams WorkspaceEdit
+  deriving (Generic)
   deriving anyclass (FromJSON, ToJSON)
 
 -- | The actual command handler
-runImportCommand :: CommandFunction ImportCommandParams
-runImportCommand _lspFuncs _state (ImportCommandParams edit) = do
-    return (Right Null, Just (WorkspaceApplyEdit, ApplyWorkspaceEditParams edit))
-
+runImportCommand :: CommandFunction IdeState ImportCommandParams
+runImportCommand _ (ImportCommandParams edit) = do
+  -- This command simply triggers a workspace edit!
+  _ <- sendRequest SWorkspaceApplyEdit (ApplyWorkspaceEditParams Nothing edit) (\_ -> pure ())
+  return (Right Null)
 ```
 
+`runImportCommand` [sends a request](https://hackage.haskell.org/package/lsp/docs/Language-LSP-Server.html#v:sendRequest) to the client using the method `SWorkspaceApplyEdit` and the parameters `ApplyWorkspaceEditParams Nothing edit`, providing a response handler that does nothing. It then returns `Right Null`, which is an empty `Aeson.Value` wrapped in `Right`.
+
 ### The code lens provider
 
 The code lens provider implements all the steps of the algorithm described earlier:
 
->  1. Request the type checking artefacts from the `ghcide` subsystem
->  2. Extract the actual import lists from the type-checked AST,
->  3. Ask GHC to produce the minimal import lists for this AST,
->  4. For every import statement without an explicit import list, find out the minimal import list, and produce a code lens to display it together with a command to graft it on.
+>  1. Request the type checking artifacts.
+>  2. Extract the actual import lists from the type-checked AST.
+>  3. Ask GHC to produce the minimal import lists for this AST.
+>  4. For each import statement lacking an explicit list, determine its minimal import list and generate a code lens displaying this list along with a command to insert it.
 
-The provider takes the usual `LspFuncs` and `IdeState` argument, as well as a `CodeLensParams` value containing the URI
-for a file, and returns an IO action producing either an error or a list of code lenses for that file.
+The provider takes the usual `LspFuncs` and `IdeState` arguments, as well as a `CodeLensParams` value containing a file URI. It returns an IO action that produces either an error or a list of code lenses for that file.
```

Comment:
`LspFuncs` does not exist anymore, they're now part of the plugin monad.

Reply:
Here I do not know what to do. This seems like an integral part of the tutorial. Should we just leave it like that at the moment and add an issue to fix the tutorial later?

Reply:
We are updating the tutorial right now, I don't think it makes sense to update the tutorial to still contain outdated info.

Reply:
Also, there are exactly two mentions of `LspFuncs`, why is it integral? Let's just delete the reference here and in the `provider` function just below it. The type `CodeLensProvider` doesn't exist any more, so that needs to be updated anyway, right?